### ISSUE TITLE
RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by jakelacey2012** [jagi/meteor-astronomy#513]
I'm getting a problem in where if I upgrade from 2.1.0 I get a call stack error.

```
W20161012-15:49:43.074(1)? (STDERR) /home/jlacey/.meteor/packages/meteor-tool/.1.3.5_1.1bjykex++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/server-lib/node_modules/fibers/future.js:280
W20161012-15:49:43.074(1)? (STDERR)                         throw(ex);
W20161012-15:49:43.074(1)? (STDERR)                               ^
W20161012-15:49:43.074(1)? (STDERR) RangeError: Maximum call stack size exceeded
```

I've simplified the problem I'm having into these model below.

``` javascript
AS = {};

AS.Object = Astro.Class.create({
    name:  'AS.Object',
});

AS.Object.extend({
    fields: {
    /**
     * @NOTE
     * Commenting out these fields, will stop the call stack error
     */
        attachments: {
            type: [AS.Object],
            optional: true
        },
        author: {
            type: AS.Object,
            optional: true
        },
    },
});

/**
 * @NOTE
 * Commenting the activity class, will stop the call stack error
 */
AS.Activity = Astro.Class.create({
    name: 'AS.Activity',
  /**
   * @NOTE
   * Commenting the collection will stop the call stack error.
   */
    collection: new Meteor.Collection('as.activities'),
    fields: {
        fieldA: {
            type: AS.Object,
        },
    },
});
```

I've created a test case which can be found here, https://github.com/Centiq/Astro-Call-Stack. 

`METEOR@1.3.5.1`, updating meteor still continues to throw the error.
- `2.1.1`, `2.1.2`, `2.1.3`, `2.1.4`, All throw the call stack error.
- However `2.1.0` works fine.

Thanks!

**Comment by lukejagodzinski** [comment]
Ah right, thanks for noticing it. The problem is nested indexes that were introduced in 2.1.1. It goes through the list of fields and for object fields through its fields and so on. It goes into infinite loop. I have to implement mechanism of detecting loops. I will try to prepare some fix today.

**Comment by jakelacey2012** [comment]
Ahh I see, thank you for taking a look at this.

I've had a look at the code you've described and I've made this change 

``` javascript
_.each(fields, (field) => {
    if (field.type.class !== Class) { // <!----- here 
        _.extend(
            indexes,
            prefixIndexes(field.type.class.definition.indexes, field.name),
            prefixIndexes(collectNestedIndexes(field.type.class), field.name)
        );
    }
});
```

I no longer get the call stack error, but I don't know how viable this fix is because I don't know how the whole indexes looping works (I've not had a look at that). I'm assuming because your already checking the fields of that class you don't need to have a look inside it.

**Comment by lukejagodzinski** [comment]
In your case such a fix solves problem but it's not universal. It would still be vulnerable. Let's take an example:

```
A class
  field: B class

B class
  field: A class
```

The `A` class has a field of type `B` which is a class, so your check would not apply here. And when checking fields of the `B` class we would encounter a field of the `A` type and we would have a loop. The check I have to make is something like stack. Adding all the classes that were already checked onto the stack. If a given class is already on the stack then do not check that class. I think it should solve a problem.

**Comment by lukejagodzinski** [comment]
Fixed and published in 2.1.5
